File: benchmark/comparative/benchmarks-QuTiP/wigner_state.py
import qutip as qt
import numpy as np
import benchmarkutils

name = "wigner_state"

samples = 3
evals = 10
cutoffs = range(50, 501, 50)


def setup(N):
    psi = qt.coherent(N, 3.0)
    xvec = np.linspace(-5, 5, 50)
    yvec = np.linspace(-5, 5, 50)
    return psi, xvec, yvec


def f(psi, xvec, yvec):
    return qt.wigner(psi, xvec, yvec)


print("Benchmarking:", name)
print("Cutoff: ", end="", flush=True)
checks = {}
results = []
for N in cutoffs:
    print(N, "", end="", flush=True)
    psi, xvec, yvec = setup(N)
    checks[N] = np.sum(f(psi, xvec, yvec))
    t = benchmarkutils.run_benchmark(f, psi, xvec, yvec, samples=samples, evals=evals)
    results.append({"N": N, "t": t})
print()

benchmarkutils.check(name, checks)
benchmarkutils.save(name, results)

File: benchmark/comparative/benchmarks-QuantumToolbox.jl/wigner_state.jl
using QuantumToolbox
using BenchmarkTools
include("benchmarkutils.jl")

name = "wigner_state"

samples = 3
evals = 10
cutoffs = [50:50:500;]

function setup(N)
    psi = coherent(N, 3.0)
    xvec = range(-5, 5, length=50)
    psi, xvec
end

function f(psi, xvec)
    wigner(psi, xvec)
end

println("Benchmarking: ", name)
print("Cutoff: ")
checks = Dict{Int, Float64}()
results = []
for N in cutoffs
    print(N, " ")
    psi, xvec = setup(N)
    w, _, _ = f(psi, xvec)
    checks[N] = sum(w)
    t = @belapsed f($psi, $xvec) samples=samples evals=evals
    push!(results, Dict("N"=>N, "t"=>t))
end
println()

benchmarkutils.check(name, checks)
benchmarkutils.save(name, results)
